### PR TITLE
feat(tools): add browser management and code interpreter lifecycle to AWS Bedrock AgentCore

### PR DESCRIPTION
**Pull request by kevin-orellana**
## Description

Add browser lifecycle management and code interpreter lifecycle tools to the `llama-index-tools-aws-bedrock-agentcore` integration package (v0.2.0), plus example notebooks for both toolkits.

**Browser tools:**
- Lifecycle management: `list_browsers`, `create_browser`, `delete_browser`, `get_browser`
- Session control: `take_control`, `release_control`, `generate_live_view_url`
- VPC support via optional `subnet_ids` and `security_group_ids` on `create_browser`

**Code interpreter tools:**
- Lifecycle management: `list_code_interpreters`, `create_code_interpreter`, `delete_code_interpreter`, `get_code_interpreter`
- File operations: `upload_file`, `upload_files`, `download_file`, `download_files`, `install_packages`, `clear_context`
- VPC support via optional `subnet_ids` and `security_group_ids` on `create_code_interpreter`

**Example notebooks** (`docs/examples/tools/`):
- `bedrock_agentcore_browser.ipynb` — direct tool usage, live view URLs, human-in-the-loop, and FunctionAgent integration
- `bedrock_agentcore_code_interpreter.ipynb` — code execution, file ops, package install, and FunctionAgent integration

**Other changes:**
- Async variants use `asyncio.to_thread()` for proper non-blocking delegation
- Dedicated control-plane client for lifecycle methods (avoids starting unnecessary sessions)
- Multi-thread isolation via `thread_id` parameter
- Bump version 0.1.2 → 0.2.0
- Bump `bedrock-agentcore` dependency 0.1.0 → 1.2.0
- Add `CHANGELOG.md`

## New Package?

- [ ] Yes
- [x] No

## Version Bump?

- [x] Yes
- [ ] No

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- [x] I added new unit tests to cover this change

**139 unit tests** covering all sync methods, error handling, and edge cases.
**35 async unit tests** covering async method variants.
**29 e2e integration tests** gated behind `AWS_AGENTCORE_E2E=1`, validated against live AgentCore services — all 29 passed. Tests cover:
- All sync session methods (navigate, click, extract, file ops, etc.)
- Lifecycle create/get/delete for both browser and code interpreter (gated behind `AWS_AGENTCORE_ROLE_ARN`)
- VPC configuration with real subnets and security groups (gated behind `AWS_AGENTCORE_SUBNET_IDS` + `AWS_AGENTCORE_SG_IDS`)
- Async variants via `asyncio.to_thread()` (code interpreter)
- Multi-thread isolation, batch downloads, task start/get/stop
- `take_control` / `release_control` browser session control

```bash
# Unit + async tests
uv run pytest tests/test_browser.py tests/test_code_interpreter.py tests/test_browser_async.py tests/test_code_interpreter_async.py -v

# E2E tests (requires AWS credentials with AgentCore access)
AWS_AGENTCORE_E2E=1 AWS_REGION=us-west-2 \
  AWS_AGENTCORE_ROLE_ARN=arn:aws:iam::ACCOUNT:role/ROLE \
  AWS_AGENTCORE_SUBNET_IDS=subnet-xxx,subnet-yyy \
  AWS_AGENTCORE_SG_IDS=sg-xxx \
  uv run pytest tests/test_browser_e2e.py tests/test_code_interpreter_e2e.py -v
```

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `uv run make format; uv run make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    def test_click_element(self, tool_spec):
+        tool_spec.navigate_browser("https://example.com")
+        result = tool_spec.click_element("a")
+        assert "Error" not in result, f"Expected successful click, got: {result}"
+
+    def test_navigate_back(self, tool_spec):
+        tool_spec.navigate_browser("https://example.com")
+        tool_spec.navigate_browser("https://www.iana.org/domains/reserved")
+        result = tool_spec.navigate_back()
+        assert "Navigated back" in result or "example.com" in result.lower(), (
```
[the remark this answers is not in the document]
Comment (kevin-orellana) [reply llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py]: This has been addressed

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        tool_spec.navigate_browser("https://example.com")
+        result = tool_spec.extract_text()
+        assert "Example Domain" in result, (
+            f"Expected 'Example Domain' in text, got: {result}"
+        )
+
+    def test_extract_hyperlinks(self, tool_spec):
+        tool_spec.navigate_browser("https://example.com")
+        result = tool_spec.extract_hyperlinks()
+        assert "iana.org" in result.lower(), (
```
[the remark this answers is not in the document]
Comment (kevin-orellana) [reply llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py:90]: This has been addressed

**Comment by kevin-orellana** [comment]
Hi maintainers — CI workflows are queued but waiting for first-time contributor approval. Could someone approve the runs when you get a chance? Thanks!

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        tool_spec.navigate_browser("https://example.com")
+        result = tool_spec.extract_text()
+        assert "Example Domain" in result, (
+            f"Expected 'Example Domain' in text, got: {result}"
+        )
+
+    def test_extract_hyperlinks(self, tool_spec):
+        tool_spec.navigate_browser("https://example.com")
+        result = tool_spec.extract_hyperlinks()
+        assert "www.iana.org" in result.lower(), (
```
[the remark this answers is not in the document]
Comment (AstraBert) [reply llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_browser_e2e.py]: I am pretty sure we can ignore these

**Comment by AstraBert** [comment]
There were a couple of CodeQL alerts on the code for tests, I dismissed them as they didn't seem to relevant

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_code_interpreter_e2e.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+
+        result = await async_tool_spec.aexecute_code("print(x)")
+        assert "NameError" in result, (
+            f"Expected NameError after async clearing context, got: {result}"
+        )
+
+    @pytest.mark.asyncio
+    async def test_async_lifecycle(self, async_tool_spec):
+        if not ROLE_ARN:
+            pytest.skip("AWS_AGENTCORE_ROLE_ARN not set")
```
Comment (AstraBert) [inline llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/tests/test_code_interpreter_e2e.py]: Would prefer to this to be a `@pytest.mark.skipif` decorator instead of having it within the test logic

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/README.md
Comment (sundargthb) [inline llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/README.md:1]: can we add the pre-reqs from the example documents here as well - aws credentials and IAM permissions. Also link the examples here.

File: llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/llama_index/tools/aws_bedrock_agentcore/browser/base.py [hunk 5]
```diff
@@ -711,6 +744,396 @@ async def acurrent_webpage(
...
+
+        """
+        return await asyncio.to_thread(
+            self.list_browsers,
+            browser_type=browser_type,
+            max_results=max_results,
+            thread_id=thread_id,
+        )
+
+    def create_browser(
```
Comment (logan-markewich) [inline llama-index-integrations/tools/llama-index-tools-aws-bedrock-agentcore/llama_index/tools/aws_bedrock_agentcore/browser/base.py:859]: Some of these functions don't quite make sense to me as agent tools. Would an agent be expected to call `create_browser()` or is this supposed to be more of a function that a user would invoke? If its for an agent, whats the expected flow?

**Comment by kevin-orellana** [comment]
Thanks for the reviews! Addressing all feedback in the latest push:

**@logan-markewich** - Great catch on the lifecycle methods. You're right that `create_browser`, `delete_browser`, `list_browsers`, `get_browser` (and the equivalent Code Interpreter ones) don't belong as agent tools. I've removed all 8 lifecycle methods from `spec_functions` so `to_tool_list()` no longer exposes them. They remain as regular class methods for programmatic use. Updated tool counts: Browser 10 (was 14), Code Interpreter 15 (was 19).

**@sundargthb** - Added prerequisites section (AWS credentials, IAM permissions) and linked example notebooks in the README.

**@AstraBert** - Converted inline `pytest.skip()` calls to `@pytest.mark.skipif` decorators in the e2e tests.
